rfcs/0004-pablo-distribution.adoc

```diff
@@ -0,0 +1,480 @@
+
+"Pre-defined reward calculation epoch in seconds" = t
+
+"Reward per calculation epoch" = r.t
+
+"Previous total reward pool before the current epoch" = P
+
+"Assuming there is a per epoch calculation which adds to the pool,"
+"the total reward pool for the current epoc,"
+
+P_(current) = P + rt
+
+"Reward pool shares for n stakers"
+
+= sum (s_n)
+
+"Where " n_(th) " staker share is " s_n
+
+"Existing " n_(th) " staker reward,"
+
+x _n = P_(current) . s_n / (sum s_n)
+
+------------------------------------------------
+
+"When adding a new staker n+1, existing stakers(n) reward would be,"
+
+x_n = P_(current) . s_n / (sum s_(n+1))
+
+"As this is less than what is expected above, an adjustment " delta P " to total reward pool can be made to allow realtime reward calculations,"
+
+delta P = P_(current) . s_(n+1) / (sum s_n)
+
+x_n = (P_(current) + delta P) . s_n / (sum s_(n+1))
+
+x_n = P_(current) . (1 + s_(n+1) / (sum s_n)) . s_n / (sum s_(n+1))
+
+x _n = P_(current) . s_n / (sum s_n)
+
+"**Therefore existing staker receives the same reward as before.**"
+
+
+"To compensate for this new adjustment, a reduction " d_(n + 1) " of reward for each staker needs to be tracked,"
+
+x_(n+1) = P_(current) . s_(n+1) / (sum s_(n+1)) - d_(n+1)
+
+"In general"
+
+d_n = "Any reduction in reward including already claimed"
+
+------------------------------------------------
+
+"When adding a new reward to the pool the calculations remain the same other than,"
+
+P_(current) = P + rt + P_(added)
+
+"Since already claimed rewards("d_n") are tracked for each staker, "
+
+"they can always claim the new reward share from " P_(added) " later."
+
+++++
+
+As this method uses a reward pooling based approach to calculate the rewards for each staker out of it on-demand, rest of the document refers to this as the "reward pooling(*RP*) based approach".
+
+==== Data Structures
+
+Staking rewards pallet already uses the following fNFT data structure,
+
+[source,rust]
+----
+include::../frame/composable-traits/src/staking_rewards.rs[lines=103..122]
+----
+This data structure can be modified to support the RP approach by adding/adjusting/removing some fields.
+
+[source,rust]
+----
+pub struct StakingNFT<AccountId, AssetId, Balance, Epoch, Rewards> {
+    /// Reward Pool ID from which pool to allocate rewards for this
+    pub reward_pool_id: ID,
+
+    /// The staked asset.
+    pub asset: AssetId,
+    /// The original stake this NFT was minted for.
+    /// Used for calculating the pool share.
+    pub stake: Balance,
+
+    /// New: Pool share received for this NFT
+    pub share: Balance,
+
+    /// New: reduced rewards for the fNFT (d_n)
+    pub reduction: Balance,
+
+    /// The reward epoch at which this NFT will start yielding rewards. This is no longer needed.
+    /// pub reward_epoch_start: Epoch,
+
+    /// List of reward asset/pending rewards. This is no longer needed as pending rewards are pooled
+    /// pub pending_rewards: Rewards,
```

Comment:
not sure about it.  can make nft with stake only so when it is rewarded all goes to stake. so it is locked compunding variant.

Reply:
Compounding is out of scope now as explained on slack :)  so resolving